engineio/transport/polling/encoder_test.go

```diff
@@ -0,0 +1,279 @@
+			return
+		}
+
+		if err := writer.Close(); err != nil {
+			frameError = fmt.Errorf("close frame error: %w", err)
+			return
+		}
+	}()
+
+	start := time.Now()
+	var output bytes.Buffer
+	if err := encoder.WriteFramesTo(&output); err != nil {
+		t.Fatalf("write frames to buffer error: %s", err)
+	}
+	gotWait := time.Since(start)
+
+	wg.Wait()
+	if frameError != nil {
+		t.Fatalf("write frames error: %s", frameError)
+	}
+	if math.Abs(float64(gotWait-wantWait)) >= 0.01*float64(time.Second) {
+		t.Fatalf("wait on WriteFramesTo(), want: %s, got: %s", wantWait, gotWait)
+	}
+
+	if want, got := data, output.String(); want != got {
+		t.Fatalf("output, want: %s, got: %s", want, got)
+	}
+}
+
+func TestEncoderWriteSeparatorInTextFrame(t *testing.T) {
+	var buf [100]byte
+	closed := make(chan struct{})
+
+	ping := time.Second / 10
+	encoder := newEncoder(ping, closed, buf[:])
+	defer encoder.WaitFrameClose()
+
+	writer, err := encoder.NextFrame(frame.Text)
+	if err != nil {
+		t.Fatalf("write text frame error: %s", err)
+	}
+
+	n, err := writer.Write([]byte{'a', 'b', separator, 'c', 'd'})
+	if want, got := ErrSeparatorInTextFrame, err; want != err {
+		t.Fatalf("write with separator want: %s, got: %s", want, got)
+	}
+	if want, got := 0, n; want != got {
+		t.Fatalf("write with separator, lenght want: %d, got: %d", want, got)
+	}
+
+	_, err = encoder.NextFrame(frame.Text)
+	if want, got := ErrNonCloseFrame, err; want != got {
+		t.Fatalf("write next frame error, want: %s, got: %s", want, got)
+	}
+
+	// Should able to continue writing.
+	data := "some data"
+	_, err = writer.Write([]byte(data))
+	if err != nil {
+		t.Fatalf("writer should be able to continue write, got: %s", err)
+	}
+	_ = writer.Close()
+
+	var output bytes.Buffer
+	_ = encoder.WriteFramesTo(&output)
+	if diff := cmp.Diff(data, output.String()); diff != "" {
+		t.Errorf("output diff:\n%s", diff)
+	}
+}
+
+func TestEncoderClose(t *testing.T) {
+	var buf [100]byte
+	closed := make(chan struct{})
+
+	ping := time.Second / 10
+	encoder := newEncoder(ping, closed, buf[:])
+
+	close(closed)
+	encoder.WaitFrameClose()
+
+	if _, err := encoder.NextFrame(frame.Text); err != io.EOF {
+		t.Fatalf("encoder.NextFrame() after closing should get io.EOF, got: %s", err)
+	}
+
+	var output bytes.Buffer
+	if err := encoder.WriteFramesTo(&output); err != io.EOF {
+		t.Fatalf("encoder.WriteFramesTo() after closing should get io.EOF, got: %s", err)
+	}
+
+	encoder.WaitFrameClose()
+}
+
+func TestEncoderWaitFrameClose(t *testing.T) {
+	var buf [100]byte
+	closed := make(chan struct{})
+
+	ping := time.Second / 10
+	encoder := newEncoder(ping, closed, buf[:])
+
+	wr, err := encoder.NextFrame(frame.Text)
+	if err != nil {
+		t.Fatalf("encoder.NextFrame() error: %s", err)
+	}
+
+	n, err := wr.Write([]byte("1234"))
+	if err != nil || n != 4 {
+		t.Fatalf("wr.Write() want: (4, nil), got: (%d, %s)", n, err)
+	}
+
+	wantWait := time.Second / 10
+	var wg sync.WaitGroup
+	defer wg.Wait()
```

Comment:
?

Reply:
Please say something. It's hard to guess with a question mark.